HW2/P5/physics.pyx

```diff
@@ -69,17 +74,27 @@ cdef void sub_update(FLOAT[:, ::1] XY,
     ############################################################
     # SUBPROBLEM 2: use the grid values to reduce the number of other
     # objects to check for collisions.
-    for j in range(i + 1, count):
-        XY2 = &(XY[j, 0])
-        V2 = &(V[j, 0])
-        if overlapping(XY1, XY2, R):
-            # SUBPROBLEM 4: Add locking
-            if not moving_apart(XY1, V1, XY2, V2):
-                collide(XY1, V1, XY2, V2)
-
-            # give a slight impulse to help separate them
-            for dim in range(2):
-                V2[dim] += eps * (XY2[dim] - XY1[dim])
+    XY1_gridX, XY1_gridY = int(XY[i, 0] / grid_spacing), int(XY[i, 1] / grid_spacing)
+    # for j in range(i + 1, count):
+    for grid_i in range(max(0, XY1_gridX - 3), min(Grid.shape[0] - 1, XY1_gridX + 4)):
+        for grid_j in xrange(max(0, XY1_gridY - 3), min(Grid.shape[1] - 1, XY1_gridY + 4)):
+            j = Grid[grid_i, grid_j]
+            if j <= i:
+                continue
+            acquire(locks + j)
```

Comment:
@k-ye True, if you want to do reading/writing as one atomic operation. But reading is no problem if you care about the 'write' updates (i.e. make them threadsafe). Regarding the pointer arithmetic, yes there is no semantic difference but there is an important difference in writing good, readable code. Imagine you pass this code to someone who has never done pointer arithmetic or doesn't know that this is possible in C/C++. He'll better recognize the concept locks[j]. Thus, you should use the language specific features only, when they really make sense. Also in the given case, the code somehow looks like python (which does not have a pointer concept) but you are using a specific C feature.